### PR TITLE
adding previousDay and its variations

### PR DESCRIPTION
**Pull request by LucasHFS**
[WIP] Implements #2507 request, and similar functions.

### CODE REVIEW

**Comment by tan75** [comment]
Hi @LucasHFS 
thanks a mill!
I linked the mentioned issue to this PR and marked it as WIP until it gets approval.
Can you please have a look into this test as it is failing?

```
  41 |   it('returns the previous Tuesday given the Saturday after it', function () {
    > 42 |     assert.deepStrictEqual(
```

if you prefer to run tests locally you can do 
`yarn test`

Many thanks!

**Comment by fturmel** [comment]
Minor stuff, but you have a little typo in your tests (_previours_ instead of _previous_).

Could I also propose a simpler and faster (by about 50%) implementation for `previousDay`? `nextDay` could be refactored using the same approach as well.

```typescript
export default function previousDay(dirtyDate: Date | number, day: Day): Date {
  requiredArgs(2, arguments)

  const date = toDate(dirtyDate)

  let delta = getDay(date) - day
  if (delta <= 0) delta += 7

  return subDays(date, delta)
}
```

**Comment by fturmel** [comment]
@LucasHFS @tan75 

There's already a PR for the `nextDay` optimization that has been approved (#2524). Should the change be reverted here or should I close 2524?

**Comment by tan75** [comment]
> @LucasHFS @tan75
> 
> There's already a PR for the `nextDay` optimization that has been approved (#2524). Should the change be reverted here or should I close 2524?

This PR 2524 will be included in our next release hopefully soon enough.
And https://github.com/date-fns/date-fns/pull/2522 can be added after. If any conflicts - we will fix it!

Thanks guys! 👍   🥇

**Review by tan75 — APPROVED** [review]
Thanks a lot for your contribution! 🚀

**Comment by fturmel** [comment]
Just had a look again at this PR and I think there's no need to convert the date argument using `toDate` before passing it to `previousDay`. This optimization applies to all the next* functions as well (nextMonday, etc).

ex:
```js
export default function previousThursday(date: Date | number): Date {
  requiredArgs(1, arguments)
  return previousDay(toDate(date), 4)
}
```

could simply be:
```js
export default function previousThursday(date: Date | number): Date {
  requiredArgs(1, arguments)
  return previousDay(date, 4)
}
```

**Comment by tan75** [comment]
@fturmel this totally makes sense! 
Will need to change the next* ones in a separate PR.

File: src/previousSunday/index.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * When is the previous Sunday?
+ *
+ * @param {Date | number} date - the date to start counting from
+ * @returns {Date} the previous Sunday
+ * @throws {TypeError} 1 argument required
+ *
+ * @example
+ * // When is the previous Sunday before Jun, 21, 2021?
+ * const result = previousSunday(new Date(2021, 5, 21))
+ * //=> Fri June 20 2021 00:00:00
```
Comment (kossnocorp) [inline src/previousSunday/index.ts:19]: Wrong output

File: src/previousSaturday/index.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * When is the previous Saturday?
+ *
+ * @param {Date | number} date - the date to start counting from
+ * @returns {Date} the previous Saturday
+ * @throws {TypeError} 1 argument required
+ *
+ * @example
+ * // When is the previous Saturday before Jun, 20, 2021?
+ * const result = previousSaturday(new Date(2021, 5, 20))
+ * //=> Fri June 19 2021 00:00:00
```
Comment (kossnocorp) [inline src/previousSaturday/index.ts:19]: Wrong

**Review by kossnocorp — CHANGES_REQUESTED** [review]

File: src/previousSunday/index.ts (hunk 1 above)
Comment (tan75) [reply src/previousSunday/index.ts:19]: ✅. Fixed

File: src/previousSaturday/index.ts (hunk 2 above)
Comment (tan75) [reply src/previousSaturday/index.ts]: ✅   Fixed

File: src/previousDay/index.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * @name previousDay
+ * @category Weekday Helpers
+ * @summary When is the previous day of the week?
+ *
+ * @description
+ * When is the previous day of the week? 0-6 the day of the week, 0 represents Sunday.
+ *
+ * @param {Date | number} date - the date to check
+ * @param {Day} day - day of the week
```
Comment (kossnocorp) [inline src/previousDay/index.ts:15]:
```suggestion
 * @param {number} day - day of the week
```

**Review by kossnocorp — APPROVED** [review]

**Pull request merged**